drivers/fpga/fpga_mpfs.c

```diff
@@ -0,0 +1,382 @@
+/*
+ * Copyright (c) 2024 Microchip Technology Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT microchip_mpfs_fpga
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/flash.h>
+#include <zephyr/drivers/fpga.h>
+#include <zephyr/kernel.h>
+#include <stdio.h>
+#include <string.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(fpga_mpfs);
+
+#define SPI_FLASH_DIRECTORY_OFFSET    0x00000000
+#define SPI_FLASH_GOLDEN_IMAGE_OFFSET 0x00100400
+#define SPI_FLASH_NEW_IMAGE_OFFSET    0x01500400
+#define SPI_FLASH_SECTOR_SIZE         4096
+#define SPI_FLASH_PAGE_SIZE           256
+
+#define MSS_SCBMAILBOX      ((volatile uint32_t *)(0x37020800UL))
+#define MSS_SCB_SERVICES_CR (*(volatile uint32_t *)(0x37020050UL))
+#define MSS_SCB_SERVICES_SR (*(volatile uint32_t *)(0x37020054UL))
+
+#define SCBCTRL_SERVICESCR_REQ      (0u)
+#define SCBCTRL_SERVICESCR_REQ_MASK (1u << SCBCTRL_SERVICESCR_REQ)
+
+#define SCBCTRL_SERVICESSR_BUSY      (1u)
+#define SCBCTRL_SERVICESSR_BUSY_MASK (1u << SCBCTRL_SERVICESSR_BUSY)
+
+#define SCBCTRL_SERVICESSR_STATUS      (16u)
+#define SCBCTRL_SERVICESSR_STATUS_MASK (0xFFFFu << SCBCTRL_SERVICESSR_STATUS)
```

Comment:
```suggestion
#define SCBCTRL_SERVICESSR_STATUS_MASK GENMASK(SCBCTRL_SERVICESSR_STATUS + 3, SCBCTRL_SERVICESSR_STATUS)
```

Reply:
@fkokosinski, excellent, thanks for the pointing out that macro